atoma-p2p/src/metrics.rs

```diff
@@ -0,0 +1,123 @@
+use nvml_wrapper::{
+    enum_wrappers::device::TemperatureSensor,
+    struct_wrappers::device::{MemoryInfo, Utilization},
+    Nvml,
+};
+use serde::{Deserialize, Serialize};
+use sysinfo::{Networks, System};
+use thiserror::Error;
+use tracing::instrument;
+
+/// Structure to store the usage metrics for the node
+///
+/// This data is collected from the system and the GPU
+/// to be sent across the p2p network, for efficient request routing.
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct NodeMetrics {
+    // /// The uptime of the node
+    // pub uptime: u64,
+    /// The CPU usage of the node
+    pub cpu_usage: f32,
+    /// The amount of RAM used
+    pub ram_used: u64,
+    /// The total amount of RAM in the system
+    pub ram_total: u64,
+    /// The amount of RAM used in swap
+    pub ram_swap_used: u64,
+    /// The total amount of swap memory in the system
+    pub ram_swap_total: u64,
+    /// The number of CPUs in the system
+    pub num_cpus: u32,
+    /// The total number of bytes received from the network
+    pub network_rx: u64,
+    /// The total number of bytes transmitted to the network
+    pub network_tx: u64,
+    /// The number of GPUs in the system
+    pub num_gpus: u32,
+    /// The usage metrics for each GPU
+    pub gpus: Vec<GpuMetrics>,
+}
+
+/// Structure to store the usage metrics for each GPU
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct GpuMetrics {
+    /// The utilization of the GPU
+    pub utilization: u32,
+    /// The amount of memory used by the GPU
+    pub memory_used: u64,
+    /// The total amount of memory on the GPU
+    pub memory_total: u64,
+    /// The amount of free memory on the GPU
+    pub memory_free: u64,
+    /// The percentage of time the GPU was reading or writing
+    pub percentage_time_read_write: u32,
+    /// The percentage of time the GPU was executing
+    pub percentage_time_gpu_execution: u32,
+    /// The temperature of the GPU
+    pub temperature: u32,
+    /// The power usage of the GPU
+    pub power_usage: u32,
+}
+
+/// Returns the usage metrics for the node
+#[instrument(level = "info", target = "metrics")]
+pub fn compute_usage_metrics(mut sys: System) -> Result<NodeMetrics, NodeMetricsError> {
+    let nvml = Nvml::init()?;
+
+    let device_count = nvml.device_count()?;
+    let mut gpus = Vec::new();
+    for i in 0..device_count {
+        let device = nvml.device_by_index(i)?;
+        let Utilization { gpu, memory } = device.utilization_rates()?;
+        let MemoryInfo { used, total, free } = device.memory_info()?;
+        let temperature = device.temperature(TemperatureSensor::Gpu)?;
+        let power_usage = device.power_usage()?;
+        gpus.push(GpuMetrics {
+            utilization: gpu,
```

Comment:
Why do we use the same value for `utilizations` and `percentage_time_gpu_execution` ?

Reply:
Good observation, I will keep only the `percentage_time_gpu_execution`